hbase-it/src/test/java/org/apache/hadoop/hbase/ShellExecCoproc.java

```diff
@@ -0,0 +1,149 @@
+
+import com.google.protobuf.RpcCallback;
+import com.google.protobuf.RpcController;
+import com.google.protobuf.Service;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.coprocessor.MasterCoprocessor;
+import org.apache.hadoop.hbase.coprocessor.RegionServerCoprocessor;
+import org.apache.hadoop.hbase.coprocessor.protobuf.generated.ShellExecEndpoint;
+import org.apache.hadoop.hbase.coprocessor.protobuf.generated.ShellExecEndpoint.ShellExecRequest;
+import org.apache.hadoop.hbase.coprocessor.protobuf.generated.ShellExecEndpoint.ShellExecResponse;
+import org.apache.hadoop.hbase.ipc.CoprocessorRpcUtils;
+import org.apache.hadoop.util.Shell;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.apache.hbase.thirdparty.com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+/**
+ * Receives shell commands from the client and executes them blindly.
+ */
+@InterfaceAudience.Private
+public class ShellExecCoproc
+  extends ShellExecEndpoint.ShellExecService
+  implements MasterCoprocessor, RegionServerCoprocessor {
+  private static final Logger LOG = LoggerFactory.getLogger(ShellExecCoproc.class);
+
+  public static final String ASYNC_DELAY_KEY = "hbase.it.shellexeccoproc.async.delay";
+  public static final long DEFAULT_ASYNC_DELAY = 1_000;
+
+  private final ExecutorService workerPool;
+  private Configuration conf;
+
+  public ShellExecCoproc() {
+    workerPool = Executors.newSingleThreadExecutor(
+      new ThreadFactoryBuilder()
+        .setNameFormat(ShellExecCoproc.class.getSimpleName() + "-{}")
+        .setDaemon(true)
+        .setUncaughtExceptionHandler((t, e) -> LOG.warn("Thread {} threw", t, e))
+        .build());
+  }
+
+  @Override
+  public Iterable<Service> getServices() {
+    return Collections.singletonList(this);
+  }
+
+  @Override
+  public void start(CoprocessorEnvironment env) {
+    conf = env.getConfiguration();
+  }
+
+  @Override
+  public void shellExec(
+    final RpcController controller,
+    final ShellExecRequest request,
+    final RpcCallback<ShellExecResponse> done
+  ) {
+    final Shell.ShellCommandExecutor shell = prepareShell(request, false);
+
+    ShellExecResponse.Builder builder = ShellExecResponse.newBuilder();
+    try {
+      builder = doExec(shell, builder);
+    } catch (IOException e) {
+      LOG.info("Failure launching process", e);
+      CoprocessorRpcUtils.setControllerException(controller, e);
+    }
+
+    done.run(builder.build());
+  }
+
+  @Override
+  public void shellExecAsync(
+    final RpcController controller,
+    final ShellExecRequest request,
+    final RpcCallback<ShellExecResponse> done
+  ) {
+    final Shell.ShellCommandExecutor shell = prepareShell(request, true);
```

Comment:
nit: I think in a true async mode, this should be async too :-)

Reply:
The request is async in that the response is disconnected from the caller. It's more like a "set-and-forget" kind of method invocation. We submit the work to the background thread and there's no avenue for the caller to see the results.